File: OCaml/src/DataStructures/SinglyLinkedList.ml
type 'a t =
  | Nil
  | Node of 'a * 'a t

let rec read index lst =
  if index < 0 then None
  else
    match lst with
    | Nil -> None
    | Node (head, tail) ->
        if index = 0 then Some head else read (index - 1) tail

let rec search_aux value current_index = function
  | Nil -> None
  | Node (head, tail) ->
      if head = value then Some current_index
      else search_aux value (current_index + 1) tail

let search value lst = search_aux value 0 lst

let rec append value = function
  | Nil -> Node (value, Nil)
  | Node (head, tail) -> Node (head, append value tail)

let prepend value lst = Node (value, lst)

let rec insert value index lst =
  match (lst, index = 0) with
  | Nil, false -> Nil
  | Nil, true -> Node (value, Nil)
  | Node (head, tail), false -> Node (head, insert value (index - 1) tail)
  | Node (head, tail), true -> Node (value, Node (head, tail))

let rec delete index = function
  | Nil -> Nil
  | Node (head, tail) ->
      if index = 0 then tail else Node (head, delete (index - 1) tail)

let rec length = function
  | Nil -> 0
  | Node (_, tail) -> length tail + 1

let rec display_aux lst acc =
  match lst with
  | Nil -> acc ^ "Nil"
  | Node (head, tail) ->
      let string_of_head = head |> string_of_int in
      let new_acc = acc ^ string_of_head ^ " -> " in
      display_aux tail new_acc

let display lst = print_endline (display_aux lst "")

let rec reverse_aux lst acc =
  match lst with
  | Nil -> acc
  | Node (head, tail) -> reverse_aux tail @@ Node (head, acc)

let reverse lst = reverse_aux lst Nil

let rec map fn = function
  | Nil -> Nil
  | Node (head, tail) -> Node (fn head, map fn tail)

let rec fold_left fn acc = function
  | Nil -> acc
  | Node (head, tail) -> fold_left fn (fn acc head) tail

let rec fold_right fn lst acc =
  match lst with
  | Nil -> acc
  | Node (head, tail) -> fn head (fold_right fn tail acc)

let rec filter fn = function
  | Nil -> Nil
  | Node (head, tail) ->
      let is_true = fn head in
      if is_true then Node (head, filter fn tail) else filter fn tail

File: Rust/src/data_structures/singly_linked_list.rs
pub trait List<T> {
    fn read(&self, index: usize) -> Option<&T>;
    fn size(&self) -> usize;
    fn reverse(&self) -> Self;
    fn find(&self, val: T) -> Option<usize>;
    fn map(&self, f: impl Fn(&T) -> T) -> Self;
    fn filter(&self, f: impl Fn(&T) -> bool) -> Self;
}

#[derive(Debug)]
pub enum SinglyLinkedList<T> {
    Node(T, Box<SinglyLinkedList<T>>),
    Nil,
}

use SinglyLinkedList::*;

impl<T: Copy> SinglyLinkedList<T> {
    fn reverse_helper(&self, acc: SinglyLinkedList<T>) -> SinglyLinkedList<T> {
        match self {
            Nil => acc,
            Node(head, tail) => tail.reverse_helper(Node(*head, Box::new(acc))),
        }
    }
}

impl<T: Copy + PartialEq> List<T> for SinglyLinkedList<T> {
    fn map(&self, func: impl Fn(&T) -> T) -> Self {
        match self {
            Nil => Nil,
            Node(head, tail) => Node(func(head), Box::new(tail.map(func))),
        }
    }

    fn read(&self, index: usize) -> Option<&T> {
        match (self, index == 0) {
            (Nil, _) => None,
            (Node(head, _), true) => Some(head),
            (Node(_, tail), false) => tail.read(index - 1),
        }
    }

    fn size(&self) -> usize {
        let mut current = self;
        let mut size = 0_usize;
        while let Node(_, tail) = current {
            size += 1;
            current = tail;
        }

        size
    }

    fn reverse(&self) -> Self {
        self.reverse_helper(Nil)
    }

    fn find(&self, val: T) -> Option<usize> {
        let mut current_index = 0_usize;
        let mut current_node = self;

        while let Node(head, tail) = current_node {
            if val == *head {
                return Some(current_index);
            } else {
                current_node = tail;
                current_index += 1;
            }
        }

        None
    }

    fn filter(&self, f: impl Fn(&T) -> bool) -> Self {
        match self {
            Nil => Nil,
            Node(head, tail) => {
                if f(head) {
                    Node(*head, Box::new(tail.filter(f)))
                } else {
                    tail.filter(f)
                }
            }
        }
    }

    // fn reverse(&self) -> Self {
    //     fn reverse_aux<T: Copy>(
    //         lst: &SinglyLinkedList<T>,
    //         acc: SinglyLinkedList<T>,
    //     ) -> SinglyLinkedList<T> {
    //         match lst {
    //             Nil => acc,
    //             Node(head, tail) => reverse_aux(tail, Node(head.clone(), Box::new(acc))),
    //         }
    //     }

    //     reverse_aux(self, Nil)
    // }
}
